### PR TITLE
Specification improvement and bug fix to sign_request

### PR DESCRIPTION
**Pull request by rickhanlonii**
## Overview

Reopening https://github.com/joestump/python-oauth2/pull/138 against `develop`.

### CODE REVIEW

File: oauth2/__init__.py [hunk 1]
```diff
@@ -485,8 +485,14 @@ def sign_request(self, signature_method, consumer, token):
         """Set the signature parameter to the result of sign."""
 
         if not self.is_form_encoded:
-            # according to
+            # according to
```
Comment (betterdiff) [inline oauth2/__init__.py]: * W291 trailing whitespace

File: oauth2/__init__.py [hunk 2]
```diff
@@ -485,8 +485,14 @@ def sign_request(self, signature_method, consumer, token):
         """Set the signature parameter to the result of sign."""
 
         if not self.is_form_encoded:
-            # according to
+            # according to 
             # http://oauth.googlecode.com/svn/spec/ext/body_hash/1.0/oauth-bodyhash.html
+            # section 3.2 "If the request does not have an entity body, the hash should
```
Comment (betterdiff) [inline oauth2/__init__.py]: * E501 line too long (87 > 79 characters)
* W291 trailing whitespace

File: oauth2/__init__.py [hunk 3]
```diff
@@ -485,8 +485,14 @@ def sign_request(self, signature_method, consumer, token):
...
 
         if not self.is_form_encoded:
-            # according to
+            # according to 
             # http://oauth.googlecode.com/svn/spec/ext/body_hash/1.0/oauth-bodyhash.html
+            # section 3.2 "If the request does not have an entity body, the hash should 
+            # be taken over the empty string."
+            if self.body is None:
+                self.body = ""
+
```
Comment (betterdiff) [inline oauth2/__init__.py]: * W293 blank line contains whitespace

File: tests/test_oauth.py [hunk 4]
```diff
@@ -1611,6 +1611,70 @@ def mockrequest(cl, ur, **kw):
...
+
+            req = oauth.Request.from_consumer_and_token(self.consumer, None,
+                                                        http_method='GET', http_url=uri, parameters={})
+            req.sign_request(oauth.SignatureMethod_HMAC_SHA1(), self.consumer, None)
+            expected = parse_qsl(urlparse.urlparse(req.to_url()).query)
+            actual = parse_qsl(urlparse.urlparse(ur).query)
+            self.failUnlessEqual(len(expected), len(actual))
+            actual = dict(actual)
+            for key, value in expected:
+                if key not in ('oauth_signature', 'oauth_nonce', 'oauth_timestamp'):
```
Comment (betterdiff) [inline tests/test_oauth.py]: * E501 line too long (84 > 79 characters)

File: tests/test_oauth.py [hunk 5]
```diff
@@ -1611,6 +1611,70 @@ def mockrequest(cl, ur, **kw):
         client.request(uri, 'GET')
 
     @mock.patch('httplib2.Http.request')
+    def test_url_with_query_string_body_none(self, mockHttpRequest):
+        uri = 'http://example.com/foo/bar/?show=thundercats&character=snarf'
+        client = oauth.Client(self.consumer, None)
+        random_result = random.randint(1,100)
```
Comment (betterdiff) [inline tests/test_oauth.py]: * E231 missing whitespace after ','

File: tests/test_oauth.py [hunk 6]
```diff
@@ -1611,6 +1611,70 @@ def mockrequest(cl, ur, **kw):
...
+                if key not in ('oauth_signature', 'oauth_nonce', 'oauth_timestamp'):
+                    self.failUnlessEqual(actual[key], value)
+
+            return random_result
+
+        mockHttpRequest.side_effect = mockrequest
+
+        client.request(uri, 'GET', body=None)
+
+    @mock.patch('httplib2.Http.request')
```
Comment (betterdiff) [inline tests/test_oauth.py]: * F811 redefinition of unused 'test_url_with_query_string_body_none' from line 1613

File: tests/test_oauth.py [hunk 7]
```diff
@@ -1611,6 +1611,70 @@ def mockrequest(cl, ur, **kw):
...
+            self.failUnless(cl is client)
+            self.failUnlessEqual(frozenset(kw.keys()), frozenset(['method', 'body', 'redirections', 'connection_type', 'headers']))
+            self.failUnlessEqual(kw['body'], None)
+            self.failUnlessEqual(kw['connection_type'], None)
+            self.failUnlessEqual(kw['method'], 'GET')
+            self.failUnlessEqual(kw['redirections'], httplib2.DEFAULT_MAX_REDIRECTS)
+            self.failUnless(isinstance(kw['headers'], dict))
+
+            req = oauth.Request.from_consumer_and_token(self.consumer, None,
+                                                        http_method='GET', http_url=uri, parameters={})
```
Comment (betterdiff) [inline tests/test_oauth.py]: * E501 line too long (103 > 79 characters)

File: tests/test_oauth.py [hunk 8]
```diff
@@ -1611,6 +1611,70 @@ def mockrequest(cl, ur, **kw):
...
+
+        mockHttpRequest.side_effect = mockrequest
+
+        client.request(uri, 'GET', body=None)
+
+    @mock.patch('httplib2.Http.request')
+    def test_url_with_query_string_body_none(self, mockHttpRequest):
+        uri = 'http://example.com/foo/bar/?show=thundercats&character=snarf'
+        client = oauth.Client(self.consumer, None)
+        random_result = random.randint(1,100)
```
Comment (betterdiff) [inline tests/test_oauth.py]: * E231 missing whitespace after ','

File: tests/test_oauth.py [hunk 9]
```diff
@@ -1611,6 +1611,70 @@ def mockrequest(cl, ur, **kw):
...
+
+            req = oauth.Request.from_consumer_and_token(self.consumer, None,
+                                                        http_method='GET', http_url=uri, parameters={})
+            req.sign_request(oauth.SignatureMethod_HMAC_SHA1(), self.consumer, None)
+            expected = parse_qsl(urlparse.urlparse(req.to_url()).query)
+            actual = parse_qsl(urlparse.urlparse(ur).query)
+            self.failUnlessEqual(len(expected), len(actual))
+            actual = dict(actual)
+            for key, value in expected:
+                if key not in ('oauth_signature', 'oauth_nonce', 'oauth_timestamp'):
```
Comment (betterdiff) [inline tests/test_oauth.py]: * E501 line too long (84 > 79 characters)

File: tests/test_oauth.py [hunk 10]
```diff
@@ -1611,6 +1611,70 @@ def mockrequest(cl, ur, **kw):
...
+
+    @mock.patch('httplib2.Http.request')
+    def test_url_with_query_string_body_none(self, mockHttpRequest):
+        uri = 'http://example.com/foo/bar/?show=thundercats&character=snarf'
+        client = oauth.Client(self.consumer, None)
+        random_result = random.randint(1,100)
+
+        def mockrequest(cl, ur, **kw):
+            self.failUnless(cl is client)
+            self.failUnlessEqual(frozenset(kw.keys()), frozenset(['method', 'body', 'redirections', 'connection_type', 'headers']))
```
Comment (betterdiff) [inline tests/test_oauth.py]: * E501 line too long (131 > 79 characters)

File: tests/test_oauth.py [hunk 11]
```diff
@@ -1611,6 +1611,70 @@ def mockrequest(cl, ur, **kw):
...
+            self.failUnlessEqual(kw['body'], None)
+            self.failUnlessEqual(kw['connection_type'], None)
+            self.failUnlessEqual(kw['method'], 'GET')
+            self.failUnlessEqual(kw['redirections'], httplib2.DEFAULT_MAX_REDIRECTS)
```
Comment (betterdiff) [inline tests/test_oauth.py]: * E501 line too long (84 > 79 characters)

File: tests/test_oauth.py [hunk 12]
```diff
@@ -1611,6 +1611,70 @@ def mockrequest(cl, ur, **kw):
...
```
Comment (betterdiff) [inline tests/test_oauth.py]: * E501 line too long (84 > 79 characters)

File: tests/test_oauth.py [hunk 13]
```diff
@@ -1611,6 +1611,70 @@ def mockrequest(cl, ur, **kw):
...
+            self.failUnless(isinstance(kw['headers'], dict))
+
+            req = oauth.Request.from_consumer_and_token(self.consumer, None,
+                                                        http_method='GET', http_url=uri, parameters={})
+            req.sign_request(oauth.SignatureMethod_HMAC_SHA1(), self.consumer, None)
```
Comment (betterdiff) [inline tests/test_oauth.py]: * E501 line too long (84 > 79 characters)

File: tests/test_oauth.py [hunk 14]
```diff
@@ -1611,6 +1611,70 @@ def mockrequest(cl, ur, **kw):
...
+            self.failUnless(cl is client)
+            self.failUnlessEqual(frozenset(kw.keys()), frozenset(['method', 'body', 'redirections', 'connection_type', 'headers']))
+            self.failUnlessEqual(kw['body'], None)
+            self.failUnlessEqual(kw['connection_type'], None)
+            self.failUnlessEqual(kw['method'], 'GET')
+            self.failUnlessEqual(kw['redirections'], httplib2.DEFAULT_MAX_REDIRECTS)
+            self.failUnless(isinstance(kw['headers'], dict))
+
+            req = oauth.Request.from_consumer_and_token(self.consumer, None,
+                                                        http_method='GET', http_url=uri, parameters={})
```
Comment (betterdiff) [inline tests/test_oauth.py]: * E501 line too long (103 > 79 characters)

File: tests/test_oauth.py [hunk 15]
```diff
@@ -1611,6 +1611,70 @@ def mockrequest(cl, ur, **kw):
...
+            req.sign_request(oauth.SignatureMethod_HMAC_SHA1(), self.consumer, None)
```
Comment (betterdiff) [inline tests/test_oauth.py]: * E501 line too long (84 > 79 characters)

File: tests/test_oauth.py [hunk 16]
```diff
@@ -1611,6 +1611,70 @@ def mockrequest(cl, ur, **kw):
...
 
     @mock.patch('httplib2.Http.request')
+    def test_url_with_query_string_body_none(self, mockHttpRequest):
+        uri = 'http://example.com/foo/bar/?show=thundercats&character=snarf'
+        client = oauth.Client(self.consumer, None)
+        random_result = random.randint(1,100)
+
+        def mockrequest(cl, ur, **kw):
+            self.failUnless(cl is client)
+            self.failUnlessEqual(frozenset(kw.keys()), frozenset(['method', 'body', 'redirections', 'connection_type', 'headers']))
```
Comment (betterdiff) [inline tests/test_oauth.py]: * E501 line too long (131 > 79 characters)

**Comment by TimSC** [comment]
I pulled this PR into a new fork: https://github.com/TimSC/python-oauth10a

**Pull request closed without merging**